### ISSUE TITLE
Cannot save memory during the FP

### ISSUE THREAD

**Issue by SpringWave1** [ucbrise/actnn#35]
Hi, part of my work now follows the your project and I try to quantize the activation during the forward pass. 
However, i just noticed that though i can modify the ctx.saved_tensor to be my new compressed activation, the overall memory occupation doesn't decrease even increase. 
What i found is that the original fp32 activation will not be freed and still be count as part of the cuda memory usage. I just wonder the reason behind this and seeks for a solution.

For your reference, what i did is like
```
def forward(self, input):
      qconv2d.apply(input, weight, .....)
```
Where inside the qconv2d, i did
```
input_int8, scale_inp = quantize_int8(input)
...
ctx.save_for_backward(input_int8, scale_inp, weight, bias)
```
However, using the torch.cuda.memory_allocated(0), the input & input_int8 will be both saved during the path?

Hoping for your reply.

**Comment by merrymercy** [comment]
1. Try to reproduce our results and check the difference between your code and our code
2. Make sure there is no other reference to your fp32 activation tensors, so they can be freed by the garbage collector.

**Comment by SpringWave1** [comment]
> 1. Try to reproduce our results and check the difference between your code and our code
> 2. Make sure there is no other reference to your fp32 activation tensors, so they can be freed by the garbage collector.

Yeah i can reproduce the mem results of ActNN by direct run the sample you provided. And it turns out the cuda allocated memory result exactly follows the compression rate. However, when i extract a single layer out and use the quantization equation you provided, it turns out the 8-bit version will have a higher allocated memory occupation. 

I got a result for fp32 - `quantized, scale = input, 0` as:
![image](https://user-images.githubusercontent.com/35594984/160585221-571b56ed-e3d0-44d8-883f-356646b27be2.png)
While int8 `quantized, scale, q_scale, q_min = quantize_activation(input)` or my personal int8 quantization method `quantized, scale = quantize_int8(input)` with:
![image](https://user-images.githubusercontent.com/35594984/160585250-c8d19343-873b-4831-ac44-44ae9c16f3ad.png)

Do i miss / misunderstand something in this part that makes the result wrong?

My code as follows:
```python
import torch.nn as nn
import torch  
import torch.nn.functional as F
from utils import get_memory_usage
from quant import quantize_int8, dequantize
from torch.autograd import Function

import actnn
import actnn.cpp_extension.quantization as ext_quantization
import actnn.cpp_extension.minimax as ext_minimax

def quantize_and_pack(data, bits, mn, mx):
    # Pack to bitstream
    if isinstance(bits, int):
        pack_func = ext_quantization.pack_single_precision
        print("here")
    else:
        pack_func = ext_quantization.pack_mixed_precision
    # import pdb; pdb.set_trace()
    output, scale = pack_func(data, mn, mx, bits, True)

    return output, scale

def no_scheme_compute_quantization_bits(input):
    N = input.shape[0]
    D = input.shape[1]
    input_flatten = input.view(N, -1)
    num_features = input_flatten.shape[1]
    num_pixels = num_features // D


    input_groups = input_flatten.view(-1, 256)
    mn, mx = ext_minimax.minimax(input_groups)
    # import pdb; pdb.set_trace()
    b = 8
    return input_groups.view(N, -1, 256), b, mn.view(N, -1, 1), mx.view(N, -1, 1)


def quantize_activation(input, scheme=None):

    N = input.shape[0]
    if scheme:
        input_groups, q_bits, q_min, mx = scheme.compute_quantization_bits(input)
    else:
        input_groups, q_bits, q_min, mx = no_scheme_compute_quantization_bits(input)

    q_input, q_scale = quantize_and_pack(input_groups, q_bits, q_min, mx)

    # TODO convert q_bits to int8
    if input.dtype == torch.float32:
        return q_input, q_bits, q_scale.to(torch.bfloat16), q_min.to(torch.bfloat16)
    else:
        return q_input, q_bits, q_scale, q_min

class convnd(Function):
    @staticmethod
    def run_forward(n, forward_op, ctx, input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1, scheme=None):
        # if not ctx.needs_input_grad[1]:
        #     assert not ctx.needs_input_grad[0] and not ctx.needs_input_grad[1]
        #     return F.conv2d(input, weight, bias, stride, padding, dilation, groups)
        # quantized, scale, q_scale, q_min = quantize_activation(input)
        # quantized, scale = quantize_int8(input)
        quantized, scale = input, 0
        # ctx.scheme = scheme
        
        ctx.saved = quantized, weight, bias
        ctx.other_args = (input.shape, stride, padding, dilation, groups)
        ctx.scale = scale

        torch.cuda.empty_cache()
        get_memory_usage(True)

        return forward_op(input, weight, bias, stride, padding, dilation, groups)

    @staticmethod
    def run_backward(n, ctx, grad_output, bias_reduce_dims, aug):
        # if not ctx.needs_input_grad[1]:
        #     assert not ctx.needs_input_grad[0] and not ctx.needs_input_grad[1]
        #     return None, None, None, None, None, None, None, None
        q_input_shape, stride, padding, dilation, groups = ctx.other_args
        padding = aug(padding)
        stride = aug(stride)
        dilation = aug(dilation)
        scale = ctx.scale

        quantized, weight, bias = ctx.saved
        input = dequantize(quantized, scale)
        del quantized, ctx.saved

        torch.cuda.empty_cache()

        # get_memory_usage(True)

        grad_input, grad_weight = ext_backward_func.cudnn_convolution_backward(
            input, grad_output, weight, padding, stride, dilation, groups,
            True, False, False,
            [ctx.needs_input_grad[0], ctx.needs_input_grad[1]])

        if bias is not None and ctx.needs_input_grad[2]:
            grad_bias = grad_output.sum(bias_reduce_dims)
        else:
            grad_bias = None


        return grad_input, grad_weight, grad_bias, None, None, None, None, None

class conv2d(Function):
    @staticmethod
    def forward(ctx, input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1, scheme=None):
        return convnd.run_forward(2, F.conv2d, ctx, input, weight, bias, stride, padding, dilation, groups, scheme)

    @staticmethod
    def backward(ctx, grad_output):
        return convnd.run_backward(2, ctx, grad_output, [0, 2, 3], _pair)

class QConv2d(nn.Conv2d):
    def __init__(self, in_channels, out_channels, kernel_size,
                 stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros', group=0):
        super(QConv2d, self).__init__(in_channels, out_channels, kernel_size,
                                      stride, padding, dilation, groups, bias, padding_mode)

    def forward(self, input):
        
        return conv2d.apply(input, self.weight, self.bias, self.stride,
                                self.padding, self.dilation, self.groups, None)

get_memory_usage()
qconv2d = QConv2d(3,3,7).cuda()
inp = torch.rand(32,3,224,224).cuda()
inp.required_grads = False
before_ = get_memory_usage()
loss = qconv2d(inp)
after_ = get_memory_usage()
print(
    "Act", (after_ - before_) / 1024 / 1024
)
```

Many thanks for your reply!

**Comment by SpringWave1** [comment]
FYI, i finally solved the problem by replacing all the non-gemm function to the version that provided by actnn. (ReLU, BatchNorm & Maxpool, etc.) There is a wierd phenomenon that, when using the model provided by the torchvision, torch seems to optimize its fp32 memory allocation during the forward propagation in these ops. :)
Again, greatly appreciate the help you provided!
